### PR TITLE
Snapshot Timer

### PR DESCRIPTION
**Pull request by johnsonw**
Configure the timer service when the configure-snapshot endpoint is called.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2249)

### CODE REVIEW

File: iml-api/src/graphql.rs [hunk 1]
```diff
@@ -666,22 +666,26 @@ impl MutationRoot {
...
             fsname,
             use_barrier.unwrap_or_default(),
             PgInterval::try_from(interval.0)?,
         )
-        .execute(&context.pg_pool)
+        .fetch_one(&context.pg_pool)
+        .map_ok(|x| x.id)
         .await?;
 
+        configure_snapshot_timer(id, fsname, interval.0).await?;
```
Comment (jgrund) [inline iml-api/src/graphql.rs:687]: This needs to take the barrier into account.

File: iml-api/src/timer.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    service_config: String,
+}
+
+pub async fn configure_snapshot_timer(
+    config_id: i32,
+    fsname: String,
+    interval: Duration,
+) -> Result<(), ImlApiError> {
+    let iml_cmd = format!(
+        r#"date +"%Y-%m-%dT%T%z" | xargs -I % echo "iml snapshot create {} {}-{}-%"#,
```
Comment (jgrund) [inline iml-api/src/timer.rs:18]: This should always contain a comment like:

`automatically created by IML`

Comment (johnsonw) [reply iml-api/src/timer.rs:18]: Isn't that comment inside both the timer and the service?

Comment (jgrund) [reply iml-api/src/timer.rs:18]: No, it's an actual param that needs to be passed to this cli.

Comment (johnsonw) [reply iml-api/src/timer.rs:18]: I was thinking the comment in the file. I understand, I'll add it with the -c option.

**Comment by jgrund** [comment]
This is failing a few checks and needs a rebase on latest master due to conflicts.

**Comment by mkpankov** [comment]
Looks like Python format check failed

**Comment by johnsonw** [comment]
I am able to create and destroy an interval using graphql. Once the interval is created, a corresponding service will be running that will create the snapshot when the interval is reached. Here is a short video showing that the interval is created and is listed in the systemd timers. It also shows that a new snapshot is created within the first minute.

**Comment by johnsonw** [comment]
snapshot_timer

**Comment by johnsonw** [comment]
I will be following up shortly with images from the docker testing

File: iml-api/src/graphql.rs [hunk 3]
```diff
@@ -699,8 +763,7 @@ impl MutationRoot {
     async fn create_snapshot_retention(
         context: &Context,
         fsname: String,
-        delete_num: i32,
-        delete_unit: DeleteUnit,
+        delete_when: InputDeleteWhen,
```
Comment (jgrund) [inline iml-api/src/graphql.rs:766]: This should be two params like it was previously.

File: iml-api/src/graphql.rs [hunk 4]
```diff
@@ -651,7 +709,7 @@ impl MutationRoot {
         interval: GraphQLDuration,
         use_barrier: Option<bool>,
     ) -> juniper::FieldResult<bool> {
-        sqlx::query!(
+        let id = sqlx::query!(
```
Comment (jgrund) [inline iml-api/src/graphql.rs:712]: What is returned when the we hit the `ON CONFLICT DO NOTHING` part?

If it's the same id, do we end up scheduling the same timer again?

File: iml-warp-drive/src/cache.rs [hunk 5]
```diff
@@ -18,7 +18,7 @@ use iml_wire_types::{
         EnclosureType, HealthState, JobState, JobType, MemberState, SfaController, SfaDiskDrive,
         SfaEnclosure, SfaJob, SfaPowerSupply, SfaStorageSystem, SubTargetType,
     },
-    snapshot::{DeleteUnit, SnapshotInterval, SnapshotRecord, SnapshotRetention},
+    snapshot::{DeleteUnit, DeleteWhen, SnapshotInterval, SnapshotRecord, SnapshotRetention},
```
Comment (jgrund) [inline iml-warp-drive/src/cache.rs:21]: The changes in this file should be reverted.

File: iml-wire-types/src/graphql_duration.rs [hunk 6]
```diff
@@ -2,11 +2,10 @@
 // Use of this source code is governed by a MIT-style
 // license that can be found in the LICENSE file.
 
+use humantime;
```
Comment (jgrund) [inline iml-wire-types/src/graphql_duration.rs:5]: Looks like the changes in this file should be reverted.

File: iml-wire-types/src/snapshot.rs [hunk 7]
```diff
@@ -86,8 +86,7 @@ pub const SNAPSHOT_INTERVAL_TABLE_NAME: TableName = TableName("snapshot_interval
 pub struct SnapshotRetention {
     pub id: i32,
     pub filesystem_name: String,
-    pub delete_num: i32,
-    pub delete_unit: DeleteUnit,
+    pub delete_when: DeleteWhen,
```
Comment (jgrund) [inline iml-wire-types/src/snapshot.rs:89]: The changes in this file should be reverted.

File: migrations/20200831174542_snapshot.sql [hunk 8]
```diff
@@ -10,6 +10,19 @@ CREATE TABLE IF NOT EXISTS snapshot (
   UNIQUE (filesystem_name, snapshot_name)
 );
 
+CREATE OR REPLACE FUNCTION table_update_notify() RETURNS TRIGGER AS $$
```
Comment (jgrund) [inline migrations/20200831174542_snapshot.sql:13]: Then changes in this file should be reverted.

File: iml-services/iml-snapshot/src/main.rs [hunk 9]
```diff
@@ -3,15 +3,146 @@
 // license that can be found in the LICENSE file.
 
 use futures_util::stream::TryStreamExt;
-use iml_manager_env::get_pool_limit;
+use iml_manager_env::{get_influxdb_addr, get_influxdb_metrics_db, get_pool_limit};
```
Comment (jgrund) [inline iml-services/iml-snapshot/src/main.rs:6]: This change should be in a separate patch for retention

File: iml-services/iml-snapshot/Cargo.toml [hunk 10]
```diff
@@ -13,5 +13,7 @@ iml-rabbit = {path = "../../iml-rabbit", version = "0.3"}
 iml-service-queue = {path = "../iml-service-queue", version = "0.3"}
 iml-tracing = {version = "0.2", path = "../../iml-tracing"}
 iml-wire-types = {path = "../../iml-wire-types", version = "0.3"}
+influx_db_client = {version = "0.4", default-features = false, features = ["rustls-tls"]}
```
Comment (jgrund) [inline iml-services/iml-snapshot/Cargo.toml:16]: This change should be in a separate patch for retention

File: iml-api/src/graphql.rs [hunk 11]
```diff
@@ -24,6 +29,12 @@ use std::{
 };
 use warp::Filter;
 
+#[derive(juniper::GraphQLInputObject)]
```
Comment (jgrund) [inline iml-api/src/graphql.rs:32]: This should be reverted.

File: iml-api/src/graphql.rs [hunk 12]
```diff
@@ -2,13 +2,18 @@
...
+    timer::{configure_snapshot_timer, remove_snapshot_timer},
+};
+use chrono::{DateTime, Utc};
 use futures::{TryFutureExt, TryStreamExt};
 use iml_postgres::{sqlx, sqlx::postgres::types::PgInterval, PgPool};
 use iml_rabbit::Pool;
 use iml_wire_types::{
     graphql_duration::GraphQLDuration,
-    snapshot::{DeleteUnit, Snapshot, SnapshotInterval, SnapshotRetention},
+    snapshot::{DeleteUnit, DeleteWhen, Snapshot, SnapshotInterval, SnapshotRetention},
```
Comment (jgrund) [inline iml-api/src/graphql.rs:16]: This should be reverted.

File: iml-api/src/graphql.rs [hunk 13]
```diff
@@ -417,8 +428,7 @@ impl QueryRoot {
     async fn snapshot_retention_policies(
         context: &Context,
     ) -> juniper::FieldResult<Vec<SnapshotRetention>> {
-        let xs: Vec<SnapshotRetention> = sqlx::query_as!(
-            SnapshotRetention,
+        let xs: Vec<SnapshotRetention> = sqlx::query!(
```
Comment (jgrund) [inline iml-api/src/graphql.rs:431]: This should be reverted.

File: iml-api/src/graphql.rs [hunk 14]
```diff
@@ -431,13 +441,45 @@ impl QueryRoot {
             "#
         )
         .fetch(&context.pg_pool)
+        .map_ok(|x| SnapshotRetention {
```
Comment (jgrund) [inline iml-api/src/graphql.rs:444]: This should be reverted.

File: iml-api/src/graphql.rs [hunk 15]
```diff
@@ -431,13 +441,45 @@ impl QueryRoot {
...
 }
 
+struct SnapshotIntervalName {
+    id: i32,
+    fs_name: String,
+    timestamp: DateTime<Utc>,
+}
+
+fn parse_snapshot_name(name: &str) -> Option<SnapshotIntervalName> {
+    let name_parts: Vec<&str> = name.split("-").collect();
```
Comment (jgrund) [inline iml-api/src/graphql.rs:468]: nit, could probably pattern match the slice which would probably read cleaner:

https://doc.rust-lang.org/edition-guide/rust-2018/slice-patterns.html

File: iml-api/src/graphql.rs [hunk 16]
```diff
@@ -438,6 +443,26 @@ impl QueryRoot {
...
+    id: i32,
+    fs_name: String,
+    timestamp: DateTime<Utc>,
+}
+
+fn parse_snapshot_name(name: &str) -> Option<SnapshotIntervalName> {
+    match name.split("-").collect::<Vec<&str>>().as_slice() {
+        [id, fs, ts] => match ts.parse::<DateTime<Utc>>() {
+            Ok(ts) => Some(SnapshotIntervalName {
+                id: id.parse::<i32>().expect("get id from name"),
```
Comment (jgrund) [inline iml-api/src/graphql.rs:456]: Why panic here instead of returning `None`?

File: iml-api/src/graphql.rs [hunk 17]
```diff
@@ -651,7 +692,7 @@ impl MutationRoot {
         interval: GraphQLDuration,
         use_barrier: Option<bool>,
     ) -> juniper::FieldResult<bool> {
-        sqlx::query!(
+        let maybe_id = sqlx::query!(
```
Comment (jgrund) [inline iml-api/src/graphql.rs:695]: What is returned when the we hit the ON CONFLICT DO NOTHING part?

If it's the same id, do we end up scheduling the same timer again?

File: iml-api/src/graphql.rs (hunk 16 above)
Comment (johnsonw) [reply iml-api/src/graphql.rs:456]: If it isn't in the right format shouldn't we fail fast? If we return None for the id how should we handle it?

**Comment by johnsonw** [comment]
Docker teting:

create-snapshot

```
[root@iml-timer /]# systemctl list-timers
NEXT                         LEFT     LAST                         PASSED       UNIT                         ACTIVAT
ES
Wed 2020-09-23 14:21:32 UTC  11s left n/a                          n/a          iml-snapshot-11.timer        iml-sna
pshot-11.service
Thu 2020-09-24 14:17:06 UTC  23h left Wed 2020-09-23 14:17:06 UTC  4min 14s ago systemd-tmpfiles-clean.timer systemd
-tmpfiles-clean.service

[root@iml-timer /]# systemctl cat iml-snapshot-11.timer
# /etc/systemd/system/iml-snapshot-11.timer
# Automatically created by IML

[Unit]
Description=Create snapshot on filesystem zfsmo

[Timer]
OnActiveSec=60
OnUnitActiveSec=60
AccuracySec=1us
Persistent=true

[Install]
WantedBy=timers.target

[root@iml-timer /]# systemctl cat iml-snapshot-11.service
# /etc/systemd/system/iml-snapshot-11.service
# Automatically created by IML

[Unit]
Description=Create snapshot on filesystem zfsmo


[Service]
Type=oneshot
EnvironmentFile=/var/lib/chroma/iml-settings.conf
ExecStart=/bin/bash -c "/usr/bin/date +\"%%Y-%%m-%%dT%%TZ\" | xargs -I %% /usr/bin/iml snapshot create -b -c 'automa
tically created by IML' zfsmo 11-zfsmo-%%"

chroma=# select * from snapshot_interval;
 id | filesystem_name | use_barrier | last_run | interval
----+-----------------+-------------+----------+----------
 11 | zfsmo           | t           |          | 00:01:00

```

New snapshot at the top of the list:

image

Removing the snapshot interval:

image


```
[root@iml-timer system]# ls
default.target        getty.target.wants  multi-user.target.wants
default.target.wants  iml-timer.service   system-update.target.wants

[root@iml-timer system]# systemctl list-timers
NEXT                         LEFT     LAST                         PASSED    UNIT                         ACTIVATES
Thu 2020-09-24 14:17:06 UTC  23h left Wed 2020-09-23 14:17:06 UTC  16min ago systemd-tmpfiles-clean.timer systemd-tm
pfiles-clean.service

1 timers listed.
```

File: iml-api/src/graphql.rs (hunk 17 above)
Comment (johnsonw) [reply iml-api/src/graphql.rs:695]: After adding a timer interval, I attempted to add the same interval a second time. Here are the results:

image


```
iml_iml-api.1.s5hlsqf9at8q@adm.local    | Sep 23 14:37:27.456 DEBUG hyper::proto::h1::io: read 1900 bytes
iml_iml-api.1.s5hlsqf9at8q@adm.local    | Sep 23 14:37:27.456 DEBUG hyper::proto::h1::io: parsed 17 headers
iml_iml-api.1.s5hlsqf9at8q@adm.local    | Sep 23 14:37:27.456 DEBUG hyper::proto::h1::conn: incoming body is content-length (1189 bytes)
iml_iml-api.1.s5hlsqf9at8q@adm.local    | Sep 23 14:37:27.456 DEBUG hyper::proto::h1::conn: incoming body completed
iml_iml-api.1.s5hlsqf9at8q@adm.local    | Sep 23 14:37:27.457  INFO sqlx::query: /* SQLx ping */; rows: 0, elapsed: 390.074µs
iml_iml-api.1.s5hlsqf9at8q@adm.local    | Sep 23 14:37:27.459  INFO sqlx::query: INSERT INTO snapshot_interval ( …; rows: 0, elapsed: 1.747ms
iml_iml-api.1.s5hlsqf9at8q@adm.local    |
iml_iml-api.1.s5hlsqf9at8q@adm.local    | INSERT INTO
iml_iml-api.1.s5hlsqf9at8q@adm.local    |   snapshot_interval (
iml_iml-api.1.s5hlsqf9at8q@adm.local    |     filesystem_name,
iml_iml-api.1.s5hlsqf9at8q@adm.local    |     use_barrier,
iml_iml-api.1.s5hlsqf9at8q@adm.local    |     interval
iml_iml-api.1.s5hlsqf9at8q@adm.local    |   )
iml_iml-api.1.s5hlsqf9at8q@adm.local    | VALUES
iml_iml-api.1.s5hlsqf9at8q@adm.local    |   ($1, $2, $3) ON CONFLICT (filesystem_name, interval) DO NOTHING RETURNING id
iml_iml-api.1.s5hlsqf9at8q@adm.local    |
iml_iml-api.1.s5hlsqf9at8q@adm.local    | Sep 23 14:37:27.459 DEBUG iml_api: Some(V4(10.0.4.64:34968)) "POST /graphql HTTP/1.0" 200 "Some("https://localhost:7443/graphiql")" "Some("Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.102 Safari/537.36")" 3.097607ms
iml_iml-api.1.s5hlsqf9at8q@adm.local    | Sep 23 14:37:27.459 DEBUG hyper::proto::h1::io: flushed 157 bytes
```
Notice it did not send a PUT request to the timer. The timer logs did not log anything either. 

The next step is to check the timers. There should only be one snapshot timer, not two:
```
[root@iml-timer /]# systemctl list-timers
NEXT                         LEFT     LAST                         PASSED    UNIT                         ACTIVATES
Wed 2020-09-23 14:41:38 UTC  35s left Wed 2020-09-23 14:40:38 UTC  24s ago   iml-snapshot-12.timer        iml-snapsh
ot-12.service
Thu 2020-09-24 14:17:06 UTC  23h left Wed 2020-09-23 14:17:06 UTC  23min ago systemd-tmpfiles-clean.timer systemd-tm
pfiles-clean.service

2 timers listed.
```

This shows that a second timer was not added. In contrast, I will change the interval to a different value and I would expect that it would create a second timer:
```
[root@iml-timer /]# systemctl list-timers
NEXT                         LEFT          LAST                         PASSED    UNIT                         ACTIV
ATES
Wed 2020-09-23 14:42:38 UTC  13s left      Wed 2020-09-23 14:41:38 UTC  46s ago   iml-snapshot-12.timer        iml-s
napshot-12.service
Wed 2020-09-23 14:44:17 UTC  1min 52s left n/a                          n/a       iml-snapshot-15.timer        iml-s
napshot-15.service
Thu 2020-09-24 14:17:06 UTC  23h left      Wed 2020-09-23 14:17:06 UTC  25min ago systemd-tmpfiles-clean.timer syste
md-tmpfiles-clean.service

3 timers listed.
```

File: iml-api/src/graphql.rs (hunk 4 above)
Comment (johnsonw) [reply iml-api/src/graphql.rs]: I posted my results in the comment below.

File: iml-api/src/graphql.rs (hunk 16 above)
Comment (jgrund) [reply iml-api/src/graphql.rs:456]: I think we need to assume that the pieces are as expected or it's not been sent by us. 

In other words this should return `None` here instead of panicking.

File: iml-api/src/graphql.rs (hunk 17 above)
Comment (jgrund) [reply iml-api/src/graphql.rs:695]: Thanks

**Review by jgrund — CHANGES_REQUESTED** [review]
Just a comment on the `expect` otherwise LGTM.

File: iml-api/src/graphql.rs (hunk 16 above)
Comment (johnsonw) [reply iml-api/src/graphql.rs:456]: Ok, I understand. How does this look?
```
fn parse_snapshot_name(name: &str) -> Option<SnapshotIntervalName> {
    match name.split("-").collect::<Vec<&str>>().as_slice() {
        [id, fs, ts] => match ts.parse::<DateTime<Utc>>() {
            Ok(ts) => {
                id.parse::<i32>()
                    .map(|id| {
                        SnapshotIntervalName {
                            id,
                            fs_name: fs.to_string(),
                            timestamp: ts,
                        }
                    })
                    .ok()
            },
            Err(_) => None,
        },
        _ => None,
    }
}
```

Comment (jgrund) [reply iml-api/src/graphql.rs:456]: How about this:

```
fn parse_snapshot_name(name: &str) -> Option<SnapshotIntervalName> {
    match name.split("-").collect::<Vec<&str>>().as_slice() {
        [id, fs, ts] => {
            let ts = ts.parse::<DateTime<Utc>>().ok()?;
            let id = id.parse::<i32>().ok()?;

            Some(SnapshotIntervalName {
                id,
                fs_name: fs.to_string(),
                timestamp: ts,
            })
        }
        _ => None,
    }
}
```

Comment (johnsonw) [reply iml-api/src/graphql.rs:456]: Sure 👍

**Review by jgrund — APPROVED** [review]

**Comment by johnsonw** [comment]
Testing ldiskfs with an rpm and docker install both yield the following:

```
[root@iml-timer /]# iml snapshot interval list
+----+------------+----------+-------------+----------+
| Id | Filesystem | Interval | Use Barrier | Last Run |
+====+============+==========+=============+==========+
| 3  | fs         | 1 minute | true        | --       |
+----+------------+----------+-------------+----------+

[root@iml-timer /]# systemctl list-timers
NEXT                         LEFT     LAST                         PASSED    UNIT                         ACTIVATES
Wed 2020-09-23 19:26:09 UTC  54s left Wed 2020-09-23 19:25:09 UTC  5s ago    iml-snapshot-3.timer         iml-snaps
hot-3.service
Thu 2020-09-24 18:54:51 UTC  23h left Wed 2020-09-23 18:54:51 UTC  30min ago systemd-tmpfiles-clean.timer systemd-t
mpfiles-clean.service

[root@iml-timer /]# systemctl cat iml-snapshot-3.service
# /etc/systemd/system/iml-snapshot-3.service
# Automatically created by IML

[Unit]
Description=Create snapshot on filesystem fs


[Service]
Type=oneshot
EnvironmentFile=/var/lib/chroma/iml-settings.conf
ExecStart=/bin/bash -c "/usr/bin/date +\"%%Y-%%m-%%dT%%TZ\" | xargs -I %% /usr/bin/iml snapshot create -b -c 'autom
atically created by IML' fs 3-fs-%%"

[root@iml-timer /]# iml snapshot list fs
+------------+---------------------------+---------------------------------+-----------+------------------------------+
| fs         | 3-fs-2020-09-23T19:24:09Z | Wed, 23 Sep 2020 19:24:10 +0000 | unmounted | automatically created by IML |
+------------+---------------------------+---------------------------------+-----------+------------------------------+
| fs         | 3-fs-2020-09-23T19:25:09Z | Wed, 23 Sep 2020 19:25:10 +0000 | unmounted | automatically created by IML |
+------------+---------------------------+---------------------------------+-----------+------------------------------+
| fs         | 3-fs-2020-09-23T19:26:09Z | Wed, 23 Sep 2020 19:26:10 +0000 | unmounted | automatically created by IML |
+------------+---------------------------+---------------------------------+-----------+------------------------------+
```

File: iml-api/src/graphql.rs [hunk 18]
```diff
@@ -462,6 +489,22 @@ impl MutationRoot {
         let name = name.trim();
         validate_snapshot_name(name)?;
 
+        let snapshot_interval_name = parse_snapshot_name(name);
+        if let Some(data) = snapshot_interval_name {
+            sqlx::query!(
+                r#"
+                UPDATE snapshot_interval
+                SET last_run=$1
+                WHERE id=$2 AND filesystem_name=$3
```
Comment (ip1981) [inline iml-api/src/graphql.rs:498]: The id is supposed to be unique. Why match with the FS?

File: iml-api/src/timer.rs [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+OnUnitActiveSec={}
+AccuracySec=1us
+Persistent=true
+
+[Install]
+WantedBy=timers.target
+"#,
+        fsname,
+        interval.as_secs(),
+        interval.as_secs()
```
Comment (ip1981) [inline iml-api/src/timer.rs:45]: Could use named parameters here and above in `iml_cmd`.

**Review by ip1981 — APPROVED** [review]
